src/electionguard/discrete_log.py

```diff
@@ -0,0 +1,100 @@
+# pylint: disable=global-statement
+# support for computing discrete logs, with a cache so they're never recomputed
+
+import asyncio
+from typing import Dict, Tuple
+
+from electionguard.singleton import Singleton
+
+from .group import G, ElementModP, ONE_MOD_P, mult_p, int_to_p_unchecked
+
+DLOG_CACHE = Dict[ElementModP, int]
+DLOG_MAX = 1_000_000
+
+
+def discrete_log(element: ElementModP, cache: DLOG_CACHE) -> Tuple[int, DLOG_CACHE]:
+    """
+    Computes the discrete log (base g, mod p) of the given element,
+    with internal caching of results. Should run efficiently when called
+    multiple times when the exponent is at most in the single-digit millions.
+    Performance will degrade if it's much larger.
+
+    For the best possible performance,
+    pre-compute the discrete log of a number you expect to have the biggest
+    exponent you'll ever see. After that, the cache will be fully loaded,
+    and every call will be nothing more than a dictionary lookup.
+    """
+
+    if element in cache:
+        return (cache[element], cache)
+
+    cache = compute_discrete_log_cache(element, cache)
+    return (cache[element], cache)
+
+
+async def discrete_log_async(
+    element: ElementModP,
+    cache: DLOG_CACHE,
+    mutex: asyncio.Lock = asyncio.Lock(),
+) -> Tuple[int, DLOG_CACHE]:
+    """
+    Computes the discrete log (base g, mod p) of the given element,
+    with internal caching of results. Should run efficiently when called
+    multiple times when the exponent is at most in the single-digit millions.
+    Performance will degrade if it's much larger.
+
+    Note: *this function is thread-safe*. For the best possible performance,
+    pre-compute the discrete log of a number you expect to have the biggest
+    exponent you'll ever see. After that, the cache will be fully loaded,
+    and every call will be nothing more than a dictionary lookup.
+    """
+    if element in cache:
+        return (cache[element], cache)
+
+    async with mutex:
+        if element in cache:
```

Comment:
Why is this line in here twice?

Reply:
This is exceptionally important. Before you acquire the lock, a second thread could come through and change the cache. That's why you check it again after you acquire the lock. This was a bug in my original code, which I only found after rewriting this in Kotlin.

Reply:
value can change between entry to and after the mutex is acquired